src/react/reconcilation.js

```diff
@@ -163,47 +165,27 @@ export class Reconciler {
         this.statistics.optimizedTrees++;
         return result;
       } catch (error) {
-        if (error.name === "Invariant Violation") {
-          throw error;
-        }
-
-        let message;
-        if (error instanceof ExpectedBailOut) {
-          message = `Failed to optimize React component tree for "${
-            evaluatedRootNode.name
-          }" due to an expected bail-out: ${error.message}`;
-        } else if (error instanceof AbruptCompletion) {
-          message = `Failed to optimize React component tree for "${
-            evaluatedRootNode.name
-          }" due to an exception thrown during render phase`;
-        } else if (error instanceof FatalError) {
-          message = `Failed to optimize React component tree for "${
-            evaluatedRootNode.name
-          }" due to a fatal error during evaluation: ${error.message}`;
-        } else {
-          // if we don't know what the error is, then best to rethrow
-          throw error;
-        }
-        throw new ReconcilerRenderBailOut(message, evaluatedRootNode);
+        invariant(!(error instanceof Completion), "React root render caught a Completion instead of an expected error");
+        this._handleComponentTreeRootFailure(error, evaluatedRootNode);
+        // flow belives we can get here, when it should never be possible
+        invariant(false, "renderReactComponentTree error not handled correctly");
       }
     };
 
     let effects = this.realm.wrapInGlobalEnv(() =>
-      this.realm.evaluatePure(() =>
-        // TODO: (sebmarkbage): You could use the return value of this to detect if there are any mutations on objects other
-        // than newly created ones. Then log those to the error logger. That'll help us track violations in
-        // components. :)
-        this.realm.evaluateForEffects(
-          renderComponentTree,
-          /*state*/ null,
-          `react component: ${getComponentName(this.realm, componentType)}`
-        )
+      this.realm.evaluatePure(
+        () =>
+          // TODO: (sebmarkbage): You could use the return value of this to detect if there are any mutations on objects other
+          // than newly created ones. Then log those to the error logger. That'll help us track violations in
+          // components. :)
```

Comment:
Is it time to address this? :-)

Reply:
Hehe, this PR means we can finally remove this TODO!